src/rtr/pdu.rs

```diff
@@ -780,9 +766,255 @@ impl fmt::Debug for RouterKeyInfo{
+        &self,
+        a: &mut A
+    ) -> Result<(), io::Error> {
+        a.write_all(self.fixed.as_ref()).await?;
+        a.write_all(self.providers.as_ref()).await
+    }
+
+    /// Reads a value from a reader.
+    ///
+    /// If a value with a different PDU type is received, returns an
+    /// error.
+    pub async fn read<Sock: AsyncRead + Unpin>(
+        sock: &mut Sock 
+    ) -> Result<Self, io::Error> {
+        let header = Header::read(sock).await?;
+        if header.pdu() != Self::PDU {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "PDU type mismatch when expecting ASPA PDU"
+            ))
+        }
+        Self::read_payload(header, sock).await
+    }
+
+    /// Reads only the payload part of a value from a reader.
+    pub async fn read_payload<Sock: AsyncRead + Unpin>(
+        header: Header, sock: &mut Sock
+    ) -> Result<Self, io::Error> {
+        let provider_len = match
+            header.pdu_len()?.checked_sub(mem::size_of::<AspaFixed>())
+        {
+            Some(len) => len,
+            None => {
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    "invalid length for ASPA PDU"
+                ))
+            }
+        };
+        eprintln!("{}", provider_len);
+        let mut fixed = AspaFixed { header, .. Default::default() };
+        sock.read_exact(&mut fixed.as_mut()[Header::LEN..]).await?;
+        if provider_len
+            != usize::from(
+                u16::from_be(fixed.provider_count)
+            ) * mem::size_of::<u32>()
+        {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "invalid length for ASPA PDU"
+            ))
+        }
+        let providers = ProviderAsns::read(sock, provider_len).await?;
+        Ok(Aspa { fixed, providers })
+    }
+}
+
+impl AsRef<[u8]> for AspaFixed {
+    fn as_ref(&self) -> &[u8] {
+        unsafe {
+            slice::from_raw_parts(
+                self as *const Self as *const u8,
+                mem::size_of::<Self>()
+            )
+        }
+    }
+}
+
+impl AsMut<[u8]> for AspaFixed {
+    fn as_mut(&mut self) -> &mut [u8] {
+        unsafe {
+            slice::from_raw_parts_mut(
+                self as *mut Self as *mut u8,
+                mem::size_of::<Self>()
+            )
+        }
+    }
+}
+
+
+//------------ ProviderAsns --------------------------------------------------
+
+/// The provider ASNs of an ASPA PDU.
+#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct ProviderAsns(Bytes);
+
+impl ProviderAsns {
+    /// The maximum length in octets for a valid value.
+    const MAX_LEN: usize = (u16::MAX as usize) * mem::size_of::<u32>();
+
+    /// Creates a new value from an iterator over ASNs.
+    ///
+    /// # Panics
+    ///
+    /// The function panics if the iterator produces more than 65,535 items.
```

Comment:
perhaps not new to this PR but if we panic here, do we have any assurance that we would reject objects with excessive provider sets in the first place?

I think panic'ing may be dangerous here otherwise. ASPA object of death and all.

Reply:
I added this panic analogous to the router key where a large key will lead to a panic. But I realise that that is actually different – you are unlikely to have a large key.

There’s going to be a second PR soon since I realized that using the imutable and wire-format `ProviderAsns` type from the `pdu` module in the actual `payload::Payload` is rather quite inconvenient, so there’s going to be a new type for that atop a vec of Asns with all the set goodness. This one will also include non-panic-y ways to deal with large provider sets.